engine/src/test/java/org/hibernate/validator/test/internal/constraintvalidators/hv/UniqueItemsValidatorTest.java

```diff
@@ -0,0 +1,162 @@
+
+		@UniqueItems
+		private final List<Object> validateMe;
+
+		AnnotationContainer(List<Object> validateMe) {
+			this.validateMe = validateMe;
+		}
+	}
+
+	@Test
+	public void testValidDataPasses() {
+		List<List<Object>> input = new ArrayList<>();
+		input.add( null );
+		input.add( Collections.singletonList( null ) );
+		input.add( Collections.singletonList( "" ) );
+		input.add( Collections.singletonList( "a" ) );
+		input.add( Arrays.asList( "a", "aa" ) );
+		input.add( Arrays.asList( "^", "ˆ" ) );
+		input.add( Arrays.asList( "a", "b", "c", "1", "2", "3", null ) );
+		input.add( Arrays.asList( "null", null ) );
+		input.add( Arrays.asList( "lorem", "lorem ipsum" ) );
+
+		input.add( Arrays.asList( new TestObjectNeverEqual(), new TestObjectNeverEqual() ) );
+		input.add( Collections.singletonList( new TestObjectAlwaysEqual() ) );
+		input.add( Arrays.asList( new TestObjectA( 1 ), new TestObjectA( 2 ), new TestObjectB(), new TestObjectC() ) );
+
+		for ( List<Object> value : input ) {
+			Set<ConstraintViolation<AnnotationContainer>> results = validator.validate( new AnnotationContainer( value ) );
+			assertTrue( results.isEmpty(), "Validation should have passed for " + value + ", " + results );
+		}
+	}
+
+	@Test
+	public void testInvalidDataFails() {
+		List<List<Object>> input = new ArrayList<>();
+		input.add( Arrays.asList( null, null ) );
+		input.add( Arrays.asList( "a", "a" ) );
+		input.add( Arrays.asList( "ˆ", "ˆ" ) );
+		input.add( Arrays.asList( "a", "b", "a" ) );
+		input.add( Arrays.asList( "a", "b", "c", "b" ) );
+		input.add( Arrays.asList( "*", "*" ) );
+
+		input.add( Arrays.asList( new TestObjectAlwaysEqual(), new TestObjectAlwaysEqual() ) );
+		input.add( Arrays.asList( new TestObjectA( 1 ), new TestObjectA( 2 ), new TestObjectA( 1 ) ) );
+		input.add( Arrays.asList( new TestObjectB(), new TestObjectB() ) );
+
+		for ( List<Object> value : input ) {
+			Set<ConstraintViolation<AnnotationContainer>> results = validator.validate( new AnnotationContainer( value ) );
+			assertTrue( results.size() > 0, "Validation should have failed for " + value + ", " + results );
+		}
+	}
+
+	@Test
+	public void testMessageContainsDuplicatedValue() {
+		String duplicate = "seeme";
+		List<Object> fails = Arrays.asList( duplicate, duplicate );
+		Set<ConstraintViolation<AnnotationContainer>> results = validator.validate( new AnnotationContainer( fails ) );
+		assertTrue( results.stream().anyMatch( cv -> cv.getMessage().contains( duplicate ) ) );
+	}
+
+	static class TestObjectA {
+
+		final int value;
+
+		TestObjectA(Integer value) {
+			this.value = value;
+		}
+
+		@Override
+		public boolean equals(Object o) {
+			if ( this == o ) {
+				return true;
+			}
+			if ( !( o instanceof TestObjectA ) ) {
+				return false;
+			}
+			TestObjectA that = (TestObjectA) o;
+			return value == that.value;
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hashCode( value );
+		}
+	}
+
+	static class TestObjectB {
```

Comment:
Frankly, I don't think we need these weird cases. A test on an object with a working equals/hashCode seems sufficient to me.

Reply:
The world is full of weird use cases... I'm happy to test less if you prefer, let me know :D

Reply:
Yeah, I think it's not worth it testing the weird hashCode/equals behaviors.